### PR TITLE
[HW-3] Resolve four tasks

### PR DESCRIPTION
**Pull request by DashaSledopit**

### CODE REVIEW

File: HW03/Nastja_Kireeva/Nastja_Kireeva_hw_03_runner.rb [hunk 1]
```diff
@@ -0,0 +1,22 @@
+require_relative 'Nastja_Kireeva_hw_03_t_01'
+require_relative 'Nastja_Kireeva_hw_03_t_02'
+require_relative 'Nastja_Kireeva_hw_03_t_03'
+require_relative 'Nastja_Kireeva_hw_03_t_04'
+
+logs = "10.6.246.103 - - [23/Apr/2018:20:30:39 +0300] 'POST /test/2/messages HTTP/1.1' 200 48 0.0498
+10.6.246.101 - - [23/Apr/2018:20:30:42 +0300] 'POST /test/2/run HTTP/1.1' 200 - 0.2277
+2018-04-23 20:30:42: SSL ERROR, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSL: System error: Undefined error: 0 - 0>
+10.6.246.101 - - [23/Apr/2018:20:31:39 +0300] 'POST /test/2/messages HTTP/1.1' 200 48 0.0290"
```
Comment (B0RlS) [inline HW03/Nastja_Kireeva/Nastja_Kireeva_hw_03_runner.rb:9]:
```suggestion
logs = <<~LOGS
  10.6.246.103 - - [23/Apr/2018:20:30:39 +0300] 'POST /test/2/messages HTTP/1.1' 200 48 0.0498
  10.6.246.101 - - [23/Apr/2018:20:30:42 +0300] 'POST /test/2/run HTTP/1.1' 200 - 0.2277
  2018-04-23 20:30:42: SSL ERROR, peer: 10.6.246.101, peer cert: , #<Puma::MiniSSL::SSL: System error: Undefined error: 0 - 0>
  10.6.246.101 - - [23/Apr/2018:20:31:39 +0300] 'POST /test/2/messages HTTP/1.1' 200 48 0.0290
LOGS
```

File: HW03/Nastja_Kireeva/Nastja_Kireeva_hw_03_t_01.rb [hunk 2]
```diff
@@ -0,0 +1,6 @@
+# frozen_string_literal: true
+
+def task1(logs)
+  lines = logs.split(/\n/)
+  p lines.map { |line| line.downcase.include?('error') ? line : '' }.join
```
Comment (B0RlS) [inline HW03/Nastja_Kireeva/Nastja_Kireeva_hw_03_t_01.rb:5]: image
We need to return only the first line

File: HW03/Nastja_Kireeva/Nastja_Kireeva_hw_03_t_03.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
+require 'time'
+
+TIMESTAMP = /^\d{4}-\d{2}-\d{2}\s\d{2}:\d{2}:\d{2}.\d/.freeze
+
+def task3(logs1)
```
Comment (B0RlS) [inline HW03/Nastja_Kireeva/Nastja_Kireeva_hw_03_t_03.rb:5]: Screenshot 2021-04-28 at 10 52 48
we can have a lot of events)

Comment (B0RlS) [inline HW03/Nastja_Kireeva/Nastja_Kireeva_hw_03_t_03.rb:5]:
```suggestion
def task3(logs)
```

File: HW03/Nastja_Kireeva/Nastja_Kireeva_hw_03_t_04.rb [hunk 4]
```diff
@@ -0,0 +1,6 @@
+def task4(string)
+  letters = string.count('a-zA-Z')
+  digits = string.count('0-9')
+  count_characters = { letters: letters, digits: digits }
+  p "#{string}  — #{count_characters}"
+end
```
Comment (vladhilko) [inline HW03/Nastja_Kireeva/Nastja_Kireeva_hw_03_t_04.rb:6]: We shouldn't keep output logic inside the method ( you break single-responsibility principle ). Could you move it outside?
```suggestion
  "#{string}  — #{count_characters}"
end

p task4
```

File: HW03/Nastja_Kireeva/Nastja_Kireeva_hw_03_t_03.rb [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+
+def task3(logs)
+  lines = logs.split(/\n/)
+  events = []
+  lines.map { |line| events << line[TIMESTAMP] if line.include?('Calling') }
+  if events.size <= 1
+    p '0'
+  else
+    calculate_diff(events)
+  end
```
Comment (vladhilko) [inline HW03/Nastja_Kireeva/Nastja_Kireeva_hw_03_t_03.rb:12]: One line is preferable if possible
```suggestion
  events.empty? ? 0 : calculate_diff(events)
```

File: HW03/Nastja_Kireeva/Nastja_Kireeva_hw_03_t_03.rb [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+  events = lines.select { |line| line.include?('Calling') }.map { |line| line[TIMESTAMP] }
+  events.empty? ? 0 : calculate_diff(events)
+end
+
+def calculate_diff(events)
+  diff = []
+  events.each_cons(2) do |previous_event, next_event|
+    diff << Time.parse(next_event) - Time.parse(previous_event)
+  end
+  diff.join(',')
```
Comment (vladhilko) [inline HW03/Nastja_Kireeva/Nastja_Kireeva_hw_03_t_03.rb:15]: The same
```suggestion
  events.each_cons(2).map do |previous_event, next_event|
    Time.parse(next_event) - Time.parse(previous_event)
  end.join(',')
```

**Review by vladhilko — APPROVED** [review]

File: HW03/Nastja_Kireeva/Nastja_Kireeva_hw_03_t_02.rb [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+  lines = logs.split(/\n/)
+
+  lines.select { |line| line.include?('POST') }.map do |line|
+    datetime = line.scan(DATETIME).first
+    address = line.scan(ADDRESS).first.upcase
+    ip = line.scan(IP).first
+    post_requests << "#{datetime} FROM: #{ip} TO: #{address}"
+  end
+
+  post_requests
```
Comment (vladhilko) [inline HW03/Nastja_Kireeva/Nastja_Kireeva_hw_03_t_02.rb:16]: We can remove temp variable 😉 
```suggestion
  lines.select { |line| line.include?('POST') }.map do |line|
    datetime = line.scan(DATETIME).first
    address = line.scan(ADDRESS).first.upcase
    ip = line.scan(IP).first
    "#{datetime} FROM: #{ip} TO: #{address}"
  end
```

**Pull request merged**